### ISSUE TITLE
Inconsistent result betwen gsDesign and rpact/EAST when randomization ratio is not 1:1

### ISSUE THREAD

**Issue by wyf88** [keaven/gsDesign#6]
Dear Keaven, 

When I compared the results of group sequential design with gsDesign, rpact and EAST, I found that gsDesign gives different results from rpact and EAST when the randomization ratio is not 1:1, and I was wondering if this is because of different method implemented, or potential bug. Below is a toy example:

- Group sequential design with 2 looks (one IA, one FA) using O'Brien-Fleming type alpha-spending function, with overall one-sided significance level of 0.025 and power of 0.8. The IA is conducted at 70% information fraction
- A total sample size of 200 and a randomization ration of 2:1
- Assume a median TTE of 10 and a hazard ratio of 0.55
- Constant enrollment of 20 subjects per month in 10 months

Results with rpact are as follows:
```r
library(rpact)
design2 <- getDesignGroupSequential(
  sided = 1, alpha = 0.025, beta = 0.2,
  kMax = 2, informationRates = c(0.7, 1),
  typeOfDesign = "asOF")
design2
```
```
Design parameters and output of group sequential design:

User defined parameters:
  Type of design                               : asOF 
  Stages                                       : 1, 2 
  Information rates                            : 0.700, 1.000 

Derived from user defined parameters:
  Maximum number of stages                     : 2 

Default parameters:
  Significance level                           : 0.0250 
  Type II error rate                           : 0.2 
  Two-sided power                              : FALSE 
  Test                                         : one-sided 
  Tolerance                                    : 1e-08 
  Type of beta spending                        : none 

Output:
  Cumulative alpha spending                    : 0.007384, 0.025000 
  Critical values                              : 2.438, 2.000 
  Stage levels                                 : 0.007384, 0.022754 
```
```r
sampleSizeResults2 <- getSampleSizeSurvival(
  design = design2, typeOfComputation = "Schoenfeld", thetaH0 = 1,
  maxNumberOfSubjects = 200, allocationRatioPlanned = 2,
  lambda2 = log(2) / 10, hazardRatio = 0.55,
  accrualTime = 0, accrualIntensity = 20)
summary(sampleSizeResults2)
```
```
Interim analysis with a maximum of 2 looks, survival endpoint (group sequential design):

Stage                                                   1        2        
Information rate                                        70%      100%     
Cumulative number of events                             70.2     100.3    
Analysis time under H1                                  14.3     20       
Cumulative alpha spent                                  0.0074   0.025    
Cumulative power                                        0.469    0.8      
One-sided local significance level                      0.0074   0.0228   
Efficacy boundary (Z-value scale)                       2.438    2        
Efficacy boundary (approximate treatment effect scale)  0.539    0.655    
Exit probability for efficacy (under H1)                0.469    0.331    
Exit probability for efficacy (under H0)                0.00738  0.01762  
```
Results with gsDesign are as follows:
```r
library(gsDesign)
gs2 <- gsSurv(test.type = 1, sided = 1, alpha = 0.025, beta = 0.2, ratio = 2,
              k = 2, timing = c(0.7, 1), sfu = "OF", 
              lambdaC = log(2) / 10, hr = 0.55,
              R = 10, gamma = 20)
gsBoundSummary(gs2)
```
```
   Analysis               Value Efficacy
  IA 1: 70%                   Z   2.4000
     N: 201         p (1-sided)   0.0082
 Events: 68         HR at bound   0.5379
  Month: 14    P(Cross) if HR=1   0.0082
            P(Cross) if HR=0.55   0.4861
      Final                   Z   2.0080
     N: 201         p (1-sided)   0.0223
 Events: 97         HR at bound   0.6479
  Month: 19    P(Cross) if HR=1   0.0250
            P(Cross) if HR=0.55   0.8000
```
As you can see, gsDesign gives the number of events of 68 and 97, whereas rpact gives the number of events of 70.2 and 100.3. EAST gives 71 and 101 which are consistent with rpact. 

When the randomization ratio is changed from 2:1 to 1:1, gsDesign gives the same results as rpact.
```r
library(rpact)
design1 <- getDesignGroupSequential(
  sided = 1, alpha = 0.025, beta = 0.2,
  kMax = 2, informationRates = c(0.7, 1),
  typeOfDesign = "asOF")
sampleSizeResults1 <- getSampleSizeSurvival(
  design = design1, typeOfComputation = "Schoenfeld", thetaH0 = 1,
  maxNumberOfSubjects = 200, allocationRatioPlanned = 1,
  lambda2 = log(2) / 10, hazardRatio = 0.55,
  accrualTime = 0, accrualIntensity = 20)
summary(sampleSizeResults1)
```
```
Interim analysis with a maximum of 2 looks, survival endpoint (group sequential design):

Stage                                                   1        2        
Information rate                                        70%      100%     
Cumulative number of events                             62.4     89.2     
Analysis time under H1                                  12.3     16.5     
Cumulative alpha spent                                  0.0074   0.025    
Cumulative power                                        0.469    0.8      
One-sided local significance level                      0.0074   0.0228   
Efficacy boundary (Z-value scale)                       2.438    2        
Efficacy boundary (approximate treatment effect scale)  0.539    0.655    
Exit probability for efficacy (under H1)                0.469    0.331    
Exit probability for efficacy (under H0)                0.00738  0.01762  
```
```r
library(gsDesign)
gs1 <- gsSurv(test.type = 1, sided = 1, alpha = 0.025, beta = 0.2, ratio = 1,
              k = 2, timing = c(0.7, 1), sfu = "OF", 
              lambdaC = log(2) / 10, hr = 0.55,
              R = 10, gamma = 20)
gsBoundSummary(gs1)
```
```
   Analysis               Value Efficacy
  IA 1: 70%                   Z   2.4000
     N: 200         p (1-sided)   0.0082
 Events: 63         HR at bound   0.5453
  Month: 12    P(Cross) if HR=1   0.0082
            P(Cross) if HR=0.55   0.4861
      Final                   Z   2.0080
     N: 200         p (1-sided)   0.0223
 Events: 90         HR at bound   0.6541
  Month: 17    P(Cross) if HR=1   0.0250
            P(Cross) if HR=0.55   0.8000
```
The number of events are 62.4 and 89.2 by rpact, and 63 and 90 by gsDesign.

I find from the rpact help that "The formula of Kim & Tsiatis (Biometrics, 1990) is used to calculate the expected number of events under the alternative (see also Lakatos & Lan, Statistics in Medicine, 1992)". It seems that EAST is using the same methodology, but I'm not sure if gsDesign is based on the same method or not. It will be great if you can have a look at the issue. Thank you!

**Comment by qzhao87** [comment]
Number of events is rounded to integers in gsDesign. Seems fine to me.

**Comment by wyf88** [comment]
@qzhao87 thanks for your reply. But in my first example where the randomization ratio is not 1:1, gsDesign gives the number of events of 68 and 97, whereas rpact gives the number of events of 70.2 and 100.3, and EAST gives 71 and 101. So it seems that gsDesign gives different result from rpact and EAST. 

BTW I have updated the outputs of rpact a bit in my original post with the new summary() function to improve the readability.

**Comment by newdayqiqi** [comment]
I also found discrepancies between gsDesign and rpact/East. rpact and East always give the same results and gsDesign gives a different number. In addition, the number of events given by nEvents is the same as rpact/East, but the one given by nSurv is different and dependent on the events rate from the control arm. This can cause big difference at times. Here 6 vs 11. In another case I tried 17 vs 43, which is not ignorable.

Keaven, could you advise on the cause of the inconsistent result? Thanks,

rpact/EAST gives 5.3
> design <- getDesignGroupSequential(kMax=1, beta = 0.1)
> summary(getSampleSizeSurvival(design, thetaH0 = 1, pi2 = 0.0091, hazardRatio = 0.06,
+                               accrualTime = c(0, 24), followUpTime = 24))
Sample size calculation for a survival endpoint

Fixed sample analysis, significance level 2.5% (one-sided).
The sample size was calculated for a two-sample logrank test, 
H0: hazard ratio = 1, H1: hazard ratio = 0.06, control pi(2) = 0.009, 
event time = 12, accrual time = 24, accrual intensity = 15.4, power 90%.

Stage                               Fixed 
Efficacy boundary (z-value scale)   1.960 
Number of subjects                  370.3 
Number of events                      **5.3** 
Analysis time                        48.0 
Expected study duration              48.0 
One-sided local significance level 0.0250 
Efficacy boundary (t)               0.182


> nEvents(hr = 0.06, hr0= 1)
[1] 5.309944
> (qnorm(1-0.025)+qnorm(0.9))^2/((log(0.06))^2*0.5*0.5)
[1] 5.309944
> nSurv(lambdaC = 0.0091/12, hr = 0.06, hr0 = 1, eta = 0, T = 48,  minfup = 24, ratio =1)$d
[1] 11.21496
> nSurv(lambdaC = 0.1, hr = 0.06, hr0 = 1, eta = 0, T = 48,  minfup = 24, ratio =1)$d
[1] 5.65361

**Comment by newdayqiqi** [comment]
Some follow-up research seems to indicate nEvents is based on Schoenfeld, and nSurv is based on Lachin. Lachin may be a more accurate calculation under low hazard ratio (HR<0.2) and small number of events, which is often the case in vaccine study. If that is true, it is a unique feather for gsDesign to provide that in nSurv function, and I would hope gsDesign2 can keep that function. We only need to clarify the methodology behind it. Thank you very much!
